main/src/ca/uwaterloo/flix/language/phase/typer/EffectProvenance.scala

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2025 Gagan Chandan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.typer
+
+import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.language.ast.TypeConstructor
+import ca.uwaterloo.flix.language.ast.Symbol
+import ca.uwaterloo.flix.language.phase.typer.TypeConstraint
+import ca.uwaterloo.flix.language.phase.typer.TypeConstraint.Provenance
+import ca.uwaterloo.flix.util.InternalCompilerException
+import scala.collection.immutable.SortedSet
+
+object EffectProvenance {
+  // Whether to enable effect provenance debugging.
+  private val enableEffectProvenance: Boolean = true
+
+  // First, we find all equality constraints that have a provenance of type
+  // Provenance.Match with a concrete effect on one side and an effect variable
+  // on the other side.
+  //
+  // For each such variable and set of concrete effects , we track a path
+  // through the constraints until we reach a constraint that has a
+  // provenance of type Provenance.ExpectEffect and an effect in the
+  // in the concrete set is not in the expected effects.
+  def debug(constrs0: List[TypeConstraint]): Unit = {
+    if (enableEffectProvenance) {
+      constrs0.foreach { constr =>
+        constr match {
+          case TypeConstraint.Equality(_, _, prov) => {
+            prov match {
+              case Provenance.Source(eff1, eff2, loc) => {
+                (eff1, eff2) match {
+                  case _ => {
+                    eff2 match {
+                      case Type.Pure => findPath(eff1, Symbol.mkEffectSym("Pure"), constr :: Nil, constrs0.filterNot(_ == constr))
```

Comment:
Making the effect symbol `Pure` seems brittle.

Would it make sense instead to create an `Option`-like type?

```scala
sealed trait XYZ

object XYZ {
    case class Effect(effectSym: EffSym) extends XYZ
    case class Pure extends XYZ
}
```